### PR TITLE
listener: add listen() error handling

### PR DESCRIPTION
**Pull request by mattklein123**
Related to https://github.com/envoyproxy/envoy/issues/14784

Signed-off-by: Matt Klein <mklein@lyft.com>

Risk Level: Low
Testing: New UT
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

**Comment by mattklein123** [comment]
This has some test ASAN issues that I need to fix, but more importantly it looks like some of the integration tests on Windows are actually failing now because listen() is failing. That doesn't seem good. cc @envoyproxy/windows-dev 

/wait

**Comment by mattklein123** [comment]
I pushed a fix for ASAN but that needs more scrutiny as it seems like a real bug in the init manager code. cc @htuch.

I still need help from @davinci26 next week to fix the Windows issues.

/wait

File: source/common/init/target_impl.cc [hunk 1]
```diff
@@ -45,8 +45,14 @@ bool TargetImpl::ready() {
   if (watcher_handle_) {
     // If we have a handle for the ManagerImpl's watcher, signal it and then reset so it can't be
     // accidentally signaled again.
+    // NOTE: The still_alive_ guard is used here to avoid the scenario in which as a result of
+    // calling ready() this target is destroyed. This is possible in practice, for example when
+    // a listener is deleted as a result of a failure in the context of the ready() call.
+    std::weak_ptr<bool> weak_still_alive = still_alive_;
     const bool result = watcher_handle_->ready();
-    watcher_handle_.reset();
+    if (!weak_still_alive.expired()) {
```
Comment (lambdai) [inline source/common/init/target_impl.cc:53]: should this still_alive be added and checked by shared_target? Or we can wait until the next crash :)
```SharedTargetImpl::ready()```

Comment (mattklein123) [reply source/common/init/target_impl.cc:53]: Yeah, if we agree this is a legitimate fix, I will fix this also and add specific tests for this. Do we believe this is the right fix though? IMO it is, as this is the crashing flow I as hitting: listener starts warming -> listener warm completes and calls ready on listener manager -> listener manager does final init of listener including calling listen() -> listen() fails -> listener manager removes listener -> crash. This is all in the same call stack.

Comment (mattklein123) [reply source/common/init/target_impl.cc]: I did a slightly different fix after talking with @htuch. I also did it in both places. PTAL.

File: source/common/init/target_impl.cc [hunk 2]
```diff
@@ -45,8 +45,11 @@ bool TargetImpl::ready() {
   if (watcher_handle_) {
     // If we have a handle for the ManagerImpl's watcher, signal it and then reset so it can't be
     // accidentally signaled again.
-    const bool result = watcher_handle_->ready();
-    watcher_handle_.reset();
+    // NOTE: We must move watcher_handle_ to a local to avoid the scenario in which as a result of
+    // calling ready() this target is destroyed. This is possible in practice, for example when
+    // a listener is deleted as a result of a failure in the context of the ready() call.
+    auto local_watcher_handle = std::move(watcher_handle_);
+    const bool result = local_watcher_handle->ready();
```
[the remark this answers is not in the document]
Comment (lambdai) [reply source/common/init/target_impl.cc:52]: It is nice that  Target is the single place that need concern about deleting itself when the member function is called.

File: source/common/init/target_impl.cc (hunk 1 above)
Comment (htuch) [reply source/common/init/target_impl.cc]: New solution looks good modulo nits.

**Comment by mattklein123** [comment]
@davinci26 now that you are back from vacation are you able to help me sort out the Windows failures? I can either add a patch to this PR if you want to fix the listen calls in a different PR I can then merge that into this. LMK.

**Comment by davinci26** [comment]
I am back today and looking into it. Let me see first how big the change is, if the patch is small, I will paste it here otherwise I will push a separate PR

**Comment by mattklein123** [comment]
@davinci26 updated PTAL

**Review by davinci26 — APPROVED** [review]
thanks for adding the TODOS, I am looking into them

**Review by ggreenway — APPROVED** [review]
LGTM

**Pull request merged**